core/src/proto/types.rs

```diff
@@ -38,416 +54,1077 @@ pub enum Error {
+        self.hash(&mut hasher);
+        Signature::new(
+            &crate::types::hash::Hash(hasher.finalize().into()),
+            sec_key,
+        )
+    }
+
+    /// Extract the data from this section if possible
+    pub fn data(&self) -> Option<Data> {
+        if let Self::Data(data) = self {
+            Some(data.clone())
+        } else {
+            None
+        }
+    }
+
+    /// Extract the extra data from this section if possible
+    pub fn extra_data_sec(&self) -> Option<Code> {
+        if let Self::ExtraData(data) = self {
+            Some(data.clone())
+        } else {
+            None
+        }
+    }
+
+    /// Extract the extra data from this section if possible
+    pub fn extra_data(&self) -> Option<Vec<u8>> {
+        if let Self::ExtraData(data) = self {
+            data.code.id()
+        } else {
+            None
+        }
+    }
+
+    /// Extract the code from this section is possible
+    pub fn code_sec(&self) -> Option<Code> {
+        if let Self::Code(data) = self {
+            Some(data.clone())
+        } else {
+            None
+        }
+    }
+
+    /// Extract the code from this section is possible
+    pub fn code(&self) -> Option<Vec<u8>> {
+        if let Self::Code(data) = self {
+            data.code.id()
+        } else {
+            None
+        }
+    }
+
+    /// Extract the signature from this section if possible
+    pub fn signature(&self) -> Option<Signature> {
+        if let Self::Signature(data) = self {
+            Some(data.clone())
+        } else {
+            None
+        }
+    }
+
+    /// Extract the ciphertext from this section if possible
+    pub fn ciphertext(&self) -> Option<Ciphertext> {
+        if let Self::Ciphertext(data) = self {
+            Some(data.clone())
+        } else {
+            None
+        }
+    }
+
+    /// Extract the MASP transaction from this section if possible
+    pub fn masp_tx(&self) -> Option<Transaction> {
+        if let Self::MaspTx(data) = self {
+            Some(data.clone())
+        } else {
+            None
+        }
+    }
+
+    /// Extract the MASP builder from this section if possible
+    pub fn masp_builder(&self) -> Option<MaspBuilder> {
+        if let Self::MaspBuilder(data) = self {
+            Some(data.clone())
+        } else {
+            None
+        }
+    }
+}
+
+/// A Namada transaction header indicating where transaction subcomponents can
+/// be found
+#[derive(
+    Clone,
+    Debug,
+    BorshSerialize,
+    BorshDeserialize,
+    BorshSchema,
+    Serialize,
+    Deserialize,
+)]
+pub struct Header {
+    /// The chain which this transaction is being submitted to
     pub chain_id: ChainId,
+    /// The time at which this transaction expires
     pub expiration: Option<DateTimeUtc>,
+    /// A transaction timestamp
+    pub timestamp: DateTimeUtc,
+    /// The SHA-256 hash of the transaction's code section
+    pub code_hash: crate::types::hash::Hash,
```

Comment:
This may be a privacy concern, as mentioned by @Fraccaman

Reply:
The `code_hash` field of `Header` is salted. That is, the WASM code hash is appended to a salt, and then hashed in order to not reveal what kind of transaction is being done. See: https://github.com/anoma/namada/blob/11a272b343e1067fe2e0c7e4673f0917496ea88e/core/src/proto/types.rs#L162 . cc @Fraccaman